process_pr.py

```diff
@@ -1255,38 +1325,74 @@ def process_pr(repo_config, gh, repo, issue, dryRun, cmsbuild_user=None, force=F
-        if [x for x in commenter_categories if x in signing_categories]:
-            ctype = ""
-            selected_cats = []
-            if re.match("^([+]1|approve[d]?|sign|signed)$", first_line, re.I):
-                ctype = "+1"
-                selected_cats = commenter_categories
-            elif re.match("^([-]1|reject|rejected)$", first_line, re.I):
-                ctype = "-1"
-                selected_cats = commenter_categories
-            elif re.match("^[+-][a-z][a-z0-9-]+$", first_line, re.I):
-                category_name = first_line[1:].lower()
-                if category_name in commenter_categories:
-                    ctype = first_line[0] + "1"
-                    selected_cats = [category_name]
-            if ctype == "+1":
-                for sign in selected_cats:
-                    signatures[sign] = "approved"
-                    if (test_comment is None) and (
-                        (repository in auto_test_repo) or ("*" in auto_test_repo)
-                    ):
-                        test_comment = comment
-                    if sign == "orp":
-                        mustClose = False
-            elif ctype == "-1":
-                for sign in selected_cats:
-                    signatures[sign] = "rejected"
-                    if sign == "orp":
-                        mustClose = False
-
     # end of parsing comments section
 
+    # Get the commit cache from `already_seen` commit or technical commit
+    print("Recalculating signatures")
+    cache_comment = None
+    new_comment = False
+
+    if technical_comment:
+        cache_comment = technical_comment
+    else:
+        if already_seen:
+            cache_comment = already_seen
+        else:
+            new_comment = True
+            # if not dryRun:
+            #     cache_comment = issue.create_comment(CMSBOT_TECHNICAL_MSG + "\n")
+
+    if cache_comment:
+        print("Loading commit cache")
+        seen_commits_match = REGEX_COMMITS_SEEN.search(cache_comment.body)
+        if seen_commits_match:
+            commit_cache = loads(seen_commits_match[1])
+
+    old_body = cache_comment.body if cache_comment else CMSBOT_TECHNICAL_MSG
+    new_body = REGEX_COMMITS_SEEN.sub(
+        "<!-- commits cache: " + dumps_compact(commit_cache) + " -->", old_body
+    )
+
+    if not dryRun:
+        if new_comment:
+            issue.create_comment(new_body)
+        else:
+            cache_comment.edit(new_body)
+    else:
+        if new_comment:
+            print("DRY RUN: Creating technical comment with text")
+            print(new_body)
+        else:
+            print("DRY RUN: Updating existing comment with text")
+            print(new_body)
+
+    for commit in all_commits:
+        if commit.sha not in commit_cache:
+            commit_cache[commit.sha] = {
+                "time": int(commit.commit.committer.date.timestamp()),
+                "files": get_changed_files_in_commit(repo, commit),
+            }
+
```

Comment:
I think creating/updating the cache_comment should be done after this.

Reply:
for each commit we are making one api call also we make one api call to get all the files changes here https://github.com/cms-sw/cms-bot/pull/2110/files#diff-f37d46a488510bd0b96fd201ff189f8a704b756502df6a1cd46b117986e608a4L772 . We need need to make all files changed call? Can the logic for file->package->category be move down here (this will save us one extra pi call)

Reply:
also we should add some protection against the PRs like https://github.com/cms-sw/cmssw/pull/38975 ( where someone used the wrong branch to open PR and it contains over 1000 commits). We can not afford to waste api calls for such PRs. I think for newly only PRs if commit count is above a threshold then bot should not process commits instead it should either add a comment to inform user about it. We can add a bot command to override this if a PR really contains commits more than default threshold

Reply:
Will look into it.